### ISSUE TITLE
rework high level interface

### ISSUE THREAD

**Issue by jdkloe** [jdkloe/pybufr-ecmwf#8]
The current message iteration for reading works like this:

``` python
from pybufr_ecmwf.bufr import BUFRReader
with BUFRReader(input_bufr_file) as bufr:
    for data, names, units in bufr.messages():
        <do something>
```

(thanks to @cpaulik).

However, I also wish to support a higher level interface for writing. This includes exposing the message headers as well, and make them writable. Therefore I propose to change the interface to something like this:

``` python
from pybufr_ecmwf.bufr import BUFRReader
with BUFRReader(input_bufr_file) as bufr:
    for msg in bufr:
        data, names, units = msg.data, msg.names, msg.units
        <do something>
```

For writing I could then implement something like:

``` python
from pybufr_ecmwf.bufr import BUFRWriter
with BUFRWriter(output_bufr_file) as bufr:
    for data in array2d:
       msg = bufr.new()
       msg.headers = <some settings>
       msg.template = <choose template>
       msg.data = data
```

and they could be combined like this:

``` python
from pybufr_ecmwf.bufr import BUFRReader, BUFRWriter
with BUFRReader(input_bufr_file) as bufr_in:
    with BUFRWriter(output_bufr_file) as bufr_out:
       for msg_in in bufr_in:
           msg_out = bufr_out.newmsg()
           msg_out.template = msg_in.template
           headers = msg_in.headers
           <do something to headers>
           msg_out.headers = headers
           data = msg_in.data
           <do something to data>
           msg_out.data = data
```

So this question is especially directed to @cpaulik but also other users may respond:
would this be acceptable to you? Or do you think there is another, more convenient way to do these tasks?

**Comment by cpaulik** [comment]
Looks good to me.

**Comment by jdkloe** [comment]
a first step to implement this has just been pushed. See commit https://github.com/jdkloe/pybufr-ecmwf/commit/4e89f81eb029f125549658d74bc9c5070448e431
The reader has now this interface:

```
from pybufr_ecmwf.bufr import BUFRReader
with BUFRReader(input_bufr_file) as bufr:
    for msg in bufr:
            for msg_or_subset_data in msg:
                data = msg_or_subset_data.data
                names = msg_or_subset_data.names
                units = msg_or_subset_data.units
                ...
```

The extra loop is a dummy loop in case the data can be read as 2D array, in which case it is only executed once. However, it is needed for 2 cases:
1. in case the subset arrays may have variable length, i.e. in case of delayed replication
2. in case some items are decoded to text (i.e. when the expand_flags setting is set to True). In this case the items are returned as python lists, not as 1D numpy arrays (since the only possible datatype would then be character strings, which also seems not very convenient).

**Comment by cpaulik** [comment]
Thanks. The dummy loop actually makes the interface closer to how the data is actually stored in the file. That seems good.

About 2. Would it be possible to return a numpy array of type `object` or would it not improve anything?

**Comment by jdkloe** [comment]
thanks for pointing this out. I just learned something new. I was not aware of the fact that numpy supports arrays holding generic python objects.
I just committed a change https://github.com/jdkloe/pybufr-ecmwf/commit/a7557119615ffe1b9e296b92944d6c70790aa2f1 that implements this. It should make it a lot easier for users to get the numbers in case they get mixed numbers/text strings in a decoded message.
If you are content with this changeI'll close this issue.

**Comment by jdkloe** [comment]
oops, I should have pointed you to one commit earlier: https://github.com/jdkloe/pybufr-ecmwf/commit/b6847a0e6ae1390092bbe78607011a3c83d536e5
The one mentioned above solves a different issue.

**Comment by cpaulik** [comment]
Thanks. looks good to me.

**Comment by jdkloe** [comment]
closed with release 0.82
